### PR TITLE
[Scala 3] Fix issues with contructors within given definition

### PR DESCRIPTION
**Pull request by tgodzik**

### CODE REVIEW

**Review by kitbellew — CHANGES_REQUESTED** [review]

**Review by dos65 — COMMENTED** [review]
@tgodzik Your solution works but I was thinking if there is a way to avoid `try/catch`
Could you look at [this commit](https://github.com/dos65/scalameta/commit/a038f71dc50ad43ce31619f964143debc0ce546c) ? It passes tests (except one about positions)

**Comment by tgodzik** [comment]
> @tgodzik Your solution works but I was thinking if there is a way to avoid `try/catch`
> Could you look at [this commit](https://github.com/dos65/scalameta/commit/a038f71dc50ad43ce31619f964143debc0ce546c) ? It passes tests (except one about positions)

We could avoid try/catch, but we need to fork it for sure. Having Init seemed just simpler overall and the exception should not happen in 99% of times. In case of you solution I think we are missing the first init.

**Comment by tgodzik** [comment]
> > @tgodzik Your solution works but I was thinking if there is a way to avoid `try/catch`
> > Could you look at [this commit](https://github.com/dos65/scalameta/commit/a038f71dc50ad43ce31619f964143debc0ce546c) ? It passes tests (except one about positions)
> 
> We could avoid try/catch, but we need to fork it for sure. Having Init seemed just simpler overall and the exception should not happen in 99% of times. In case of you solution I think we are missing the first init.

Or maybe it does work. Let me take a look.

**Comment by tgodzik** [comment]
@dos65 thanks for the suggestion! It seems to work correctly :tada:

File: scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala [hunk 1]
```diff
@@ -1061,7 +1061,11 @@ object TreeSyntax {
         val pearly = r(t.early, "{ ", "; ", " } with")
         val pparents = r(t.inits, " with ")
         val derived = r(t.derives, "derives ", ", ", "")
-        val withGiven = if (t.parent.exists(_.is[Defn.Given]) && !isBodyEmpty) "with" else ""
+        val isGiven = t.parent.exists(_.is[Defn.Given])
+        val withGiven = if (t.parent.exists(_.is[Defn.Given])) {
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala]: Changed to the first option

File: scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala [hunk 2]
```diff
@@ -1061,7 +1061,11 @@ object TreeSyntax {
         val pearly = r(t.early, "{ ", "; ", " } with")
         val pparents = r(t.inits, " with ")
         val derived = r(t.derives, "derives ", ", ", "")
-        val withGiven = if (t.parent.exists(_.is[Defn.Given]) && !isBodyEmpty) "with" else ""
+        val isGiven = t.parent.exists(_.is[Defn.Given])
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:1064]: Good catch! It's now used.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 3]
```diff
@@ -4200,22 +4200,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
...
-        )
-      )
+    def declType = declInit match {
+      case Left(init) => init.tpe
+      case Right(tpe) => tpe
+    }
 
+    def parents() = {
+      val firstInit = declInit match {
+        case Left(init) => init
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Changed according to @dos65 suggestion

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 4]
```diff
@@ -4200,22 +4200,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
...
+          Right(startModType())
+      }
 
-    def parents() = {
-      val parents = ListBuffer[Init](
-        atPos(decltpe.startTokenPos, decltpe.endTokenPos)(
-          Init(decltpe, autoPos(Name.Anonymous()), Nil)
-        )
-      )
+    def declType = declInit match {
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Reverted

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 5]
```diff
@@ -4200,22 +4200,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
...
+      case Left(init) => init.tpe
+      case Right(tpe) => tpe
+    }
 
+    def parents() = {
+      val firstInit = declInit match {
+        case Left(init) => init
+        case Right(tpe) => syntaxError("given cannot constructed with this type", at = tpe.pos)
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Removed altogether.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 6]
```diff
@@ -4200,23 +4200,22 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           (anonymousName, List.empty, List.empty)
       }
 
-    val decltpe = startModType()
+    val decltype = startModType()
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:4203]: still renamed... i know tpe is weird but that's what it's called in 100 other places in scalameta

**Review by dos65 — APPROVED** [review]
:+1:

**Pull request merged**